PATCH-4: Add TC01 and TC02

yenleagest opened the pull request:
Hi @nbmhoang, please review this pull request. It’s been tested on Jenkins and resulted in failures due to issues on Agoda’s side. I was able to reproduce them manually and can confirm it’s not a scripting issue.
Here is the [drive link](https://drive.google.com/file/d/1PliTeJfJBJqReKwlFlBAFVAOx7G1Mm1c/view?usp=sharing) to the report in case you’re interested.
![jenkins](https://github.com/user-attachments/assets/aa7a8201-aeff-488f-a9f0-99a0c4b675aa)

**Issues observed:**
- TC01: One of the first 5 results displays no price.
![SearchAndSortHotelTest](https://github.com/user-attachments/assets/233909be-9811-48ab-9dc5-d40a98d1c45b)
- TC02: One of the first 5 results shows a price range that falls outside the filter range.
![SearchAndFilterHotelTest](https://github.com/user-attachments/assets/24c9a1aa-52bd-430a-86e7-b40135c0a26a)

catssaymeow committed:
Add TC01 and TC02

nbmhoang commented on src/main/java/drivers/DriverUtils.java:
```diff
@@ -2,25 +2,31 @@
         return getWebDriver();
     }
+
+    public static void switchToLatestTab() {
```
Use Selenide API please

yenleagest replied:
Done.

catssaymeow committed:
Fix review comments

yenleagest commented on src/main/java/utils/CurrencyConverter.java:
```diff
@@ -0,0 +1,35 @@
+                    return currency.conversionRate;
+                }
+            }
+            return 1; // default rate if not matched
```
Updated.

yenleagest commented on src/test/java/testcases/TestBase.java:
```diff
@@ -2,11 +2,29 @@
+            data[i][0] = dataList.get(i);
+        }
+        return data;
+    }
```
Updated.

nbmhoang commented on src/main/java/pages/agoda/HomePage.java:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    @Step("Search a hotel with given information: {location}, {duration} days from {offset} {weekday}, occupancy: {occupancy}")
+    public void searchHotel(String location, DateUtils.WeekOffset offset, DateUtils.Weekday weekday, int duration, Occupancy occupancy) {
```
What is offset param used for?

yenleagest replied:
Removed as no longer needed.

nbmhoang commented on src/main/java/pages/agoda/HomePage.java:
```diff
@@ -0,0 +1,125 @@
+
+    @Step("Get current occupancy")
+    public Occupancy getCurrentOccupancy() {
+        int rooms = Integer.parseInt(new Element(OccupancyType.ROOMS.valueSelector()).getRaw().getText().trim());
```
you should write a new method for each type

yenleagest replied:
Done.

yenleagest commented:
Hi @nbmhoang all comments have been fixed, please recheck and let me know if any issues.

nbmhoang commented on src/main/java/pages/agoda/HomePage.java:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    @Step("Perform adjustment for occupancy: {occType}, increase: {increase}")
+    public void adjustOccupancy(OccupancyType occType, boolean increase) {
```
Is this a page action? Why we need to pass the increase value?

yenleagest replied:
Removed as unneeded. Thanks for pointing it out.

nbmhoang commented on src/main/java/pages/agoda/HomePage.java:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    @Step("Select date: {date}")
+    private void selectDate(String date) {
```
date data type should be date type rather than String

yenleagest replied:
Updated.

nbmhoang commented on src/main/java/utils/CurrencyConverter.java:
```diff
@@ -0,0 +1,35 @@
+    // when running on Jenkins, agoda will display the server currency
+    @AllArgsConstructor
+    enum Currency {
+        SGD("SGD", 20000),
```
do you think that the rate is necessary here?

yenleagest replied:
As stated in the comment, this is necessary when running on Jenkins. The maximum value for other currencies is much lower than for VND (e.g., for SGD it’s 2230). If you input a min and max like 500000–1000000, it has no effect because the values exceed the allowed range.

nbmhoang commented on src/main/java/utils/DateUtils.java:
Java LocalDate already handle this for u. Use it

yenleagest replied:
Removed as no longer needed.

nbmhoang commented on src/test/java/testcases/TestBase.java:
```diff
@@ -2,11 +2,29 @@
+    @DataProvider(name = "dataByMethod")
+    public Object[][] getTestDataFromYaml(Method method) throws Exception {
+        String testCaseName = method.getName();
+        List<Map<String, String>> dataList = getTestData(testCaseName);
```
In case you **really really** need parse test data from file, I suggest you create a class then use snakeyaml to deserialize it

yenleagest replied:
Updated. Yes that way is really convenient.

nbmhoang commented on src/test/java/testcases/agoda/SearchAndFilterHotelTest.java:
Your test case is too long 😭  Don’t you think it would be better to filter the origin list using all conditions at once, then compare it with the list in the UI?

yenleagest replied:
Update.

nbmhoang commented on src/main/java/elements/Element.java:
what is this. I guess you're trying to work with relative elements/locators?

yenleagest replied:
It wraps web elements and adds some extra behavior like waiting for it to be visible before clicking or clearing the textbox before entering. Do you think its a bad practice?

yenleagest replied:
Removed as no longer needed.

catssaymeow committed:
Fix review comments

yenleagest commented:
Hi @nbmhoang All comments have been addressed. However, I’d prefer to keep the [src/main/java/utils/CurrencyConverter.java](https://github.com/yenleagest/selenium-level-3/pull/4/files/6e8520b82de1b4cc0150448031b525279b1123f0#diff-7b616f6d3b21ef5d8b77ece421b1724ac15df450170aba41a2fbcb8a85b23702) as is, for the reasons mentioned earlier. I really appreciate the time and effort you’ve put into reviewing this. Thank you!

catssaymeow committed:
Update logging

catssaymeow committed:
Zip report and enhance robustness when locating hotel attr

yenleagest commented:
@nbmhoang I will include the fix for those review comments in our latets session in this PR and will let know once its done.

catssaymeow committed:
Fix last session's review comments

catssaymeow committed:
Close google pop-up before selecting currency

yenleagest commented:
Hi @nbmhoang Ive addressed all the comments from our today session. Could you please recheck when you have a chance?
#### Change summary:
- Removed HomePageActions
- Removed CurrencyConverter
- Added Project Progress & Project Structure to README
- Moved selenide.browserSize=1920x1080 to selenide.properties

I also attached the allure report as evidence that my code still works after the update. Check it out if you're interested.
[allure-report.zip](https://github.com/user-attachments/files/20835785/allure-report.zip)

nbmhoang commented on src/main/java/pages/agoda/HomePage.java:
```diff
@@ -0,0 +1,227 @@
+    }
+
+    @Step("Click on occupancy container")
+    public void selectOccupancyContainer() {
```
Rename this method. Method names should reflect business steps rather than UI interactions.

yenleagest replied:
Updated.

nbmhoang commented on src/main/java/pages/agoda/HomePage.java:
```diff
@@ -0,0 +1,227 @@
+    }
+
+    @Step("Search a hotel with given information: {location}, {duration} days from next {targetDay}, occupancy: {occupancy}")
+    public void searchHotel(String location, DayOfWeek targetDay, int duration, Occupancy occupancy) {
```
Can we declare 2nd argument as LocalDate or other date data type?

yenleagest replied:
Yes I changed the 2nd and the 3rd to LocalDate to improve the consistency :)

nbmhoang commented on src/main/java/pages/agoda/HomePage.java:
```diff
@@ -0,0 +1,227 @@
+
+    @Step("Close Google account pop-up if present")
+    private void closeGoogleIframe() {
+        // a Google pop-up will open and overlap the menu if running in other browsers rather than Chrome
```
Did you mean this frame? Try to open the Agoda on Safari but also see this popup.
Addtionally, you can check the browser by using isChrome, isIE,... from WebDriverRunner class
![image](https://github.com/user-attachments/assets/c183a2b6-583d-42e7-a9f2-8ee6f7acdeda)

nbmhoang replied:
Or just simply check it's visible or not

yenleagest replied:
This iframe appears a bit lately, after all the components are loaded. Because of that, if we check for its visiblibity, it will result a false. Anh yes it happens with Safari, as I mentioned in the comment "other browsers rather than Chrome". I will use isChorme() to do that.

yenleagest replied:
I just learned about the exists() method. It seems like a perfect fit for situations like this.

nbmhoang commented on src/test/java/testcases/agoda/SearchAndFilterHotelTest.java:
```diff
@@ -0,0 +1,95 @@
+
+        // The price and star filtered is highlighted
+        actualPriceFilter = searchResultsPage.getPriceFilterValues();
+        softAssert.assertEquals(actualPriceFilter, inputPriceFilter, "Price filter does not match\n - Expected: %s\n - Actual: %s".formatted(inputPriceFilter, actualPriceFilter));
```
in case of assertion is failed, TestNG already logged the expected and actual data in the msg for you

yenleagest replied:
Removed them all.

nbmhoang commented on src/main/java/data/models/CardContainer.java:
```diff
@@ -0,0 +1,66 @@
+@Data
+@AllArgsConstructor
+@Slf4j
+public class CardContainer {
```
Do you think should we rename this class to "Hotel"?

yenleagest replied:
Yes I did.

nbmhoang commented on src/main/java/data/models/CardContainer.java:
```diff
@@ -0,0 +1,66 @@
+     *                      - if {@code sortedByPrice} is false and {@code priceRange} is not provided, no price filtering is applied
+     * @return the filtered list of hotels
+     */
+    public static List<CardContainer> filteredHotels(List<CardContainer> hotels,
```
This method should not place here!
Try to create HotelFilters and write method for each filter(filterByDestination, filterByRating, filterByPriceRange, filterByMinimumPrice, filterByMaximumPrice,...)

nbmhoang replied:
There are two points I want to make about my sentence:
- First, as I recall from your first commit in this PR, you're applying each filter condition separately in the test case—first filtering by star, comparing data, then filtering by price, and comparing again. This makes the test steps unnecessarily long, whereas you could filter by both star and price at the same time.
- Second, I suggest creating a separate method for each filter condition (for the upcoming HotelFilters class). This will make your code more readable and reusable. Try reading the two code blocks below. Which one would you prefer?
`CardContainer.filteredHotels(hotels, location, false, 0);`
AND
`filterByDestination(
                filterByRating(
                        filterByPriceBetween(hotels, 1_000_000, 2_000_000),
                        3),
                location
        );`
![image](https://github.com/user-attachments/assets/583e91c0-5e6f-492b-9d9d-f49be4711d1b)

nbmhoang replied:
Also, if the filter condition gets too complex, you can just write a class for it!(but not this case)
---
chơi mà xoá comment 👿

yenleagest replied:
Nhờ comment của thầy, e đã được khai sáng đang đang âm thầm update ạ 🥹

yenleagest replied:
Hey @nbmhoang, I found a new way to do this - method chaining. Looks pretty awesome now 🥳 Hope you like it. Honestly, I wouldn’t have thought of it without your suggestion. Thanks for your patience 🙇🏻‍♀️

nbmhoang reviewed (changes requested):
Please see my comments as below.
For progress tracking in README file, I suggest to use markdown: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#task-lists
- [x] Task 1
- [ ] Task 2
- [x] Task 3

catssaymeow committed:
Fix recent review comments

yenleagest commented:
Hi @nbmhoang, all comments have been fixed. Please recheck. Hope I didnt ruin your Friday night.

catssaymeow committed:
Rename method

nbmhoang reviewed (changes requested):
Thanks! PR is merged and good to go :rocket:
